fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexingScrubber.java

```diff
@@ -0,0 +1,359 @@
+
+        // index must be in readable mode
+        validateOrThrowEx(store.getIndexState(index) == IndexState.READABLE, "scrubbed index is not readable");
+
+        RangeSet rangeSet = new RangeSet(indexScrubIndexRangeSubspace(store, index));
+        AsyncIterator<Range> ranges = rangeSet.missingRanges(store.ensureContextActive(), startBytes, endBytes).iterator();
+
+        final ExecuteProperties.Builder executeProperties = ExecuteProperties.newBuilder()
+                .setIsolationLevel(IsolationLevel.SNAPSHOT)
+                .setReturnedRowLimit(getLimit() + 1); // always respectLimit in this path; +1 allows a continuation item
+        final ScanProperties scanProperties = new ScanProperties(executeProperties.build());
+
+        return ranges.onHasNext().thenCompose(hasNext -> {
+            if (Boolean.FALSE.equals(hasNext)) {
+                return AsyncUtil.READY_FALSE; // no more missing ranges - all done
+            }
+            final Range range = ranges.next();
+            final Tuple rangeStart = RangeSet.isFirstKey(range.begin) ? null : Tuple.fromBytes(range.begin);
+            final Tuple rangeEnd = RangeSet.isFinalKey(range.end) ? null : Tuple.fromBytes(range.end);
+            final TupleRange tupleRange = TupleRange.between(rangeStart, rangeEnd);
+
+            RecordCursor<FDBIndexedRecord<Message>> cursor =
+                    store.scanIndexRecords(index.getName(), IndexScanType.BY_VALUE, tupleRange, null, IndexOrphanBehavior.RETURN, scanProperties);
+
+            final AtomicBoolean hasMore = new AtomicBoolean(true);
+            final AtomicReference<RecordCursorResult<FDBIndexedRecord<Message>>> lastResult = new AtomicReference<>(RecordCursorResult.exhausted());
+            final long quota = spolicy().getQuota();
+
+            return iterateRangeOnly(store, cursor,
+                    this::deleteIndexIfDangling,
+                    lastResult, hasMore, recordsScanned)
+                    .thenApply(vignore -> hasMore.get() ?
+                                          lastResult.get().get().getIndexEntry().getKey() :
+                                          rangeEnd)
+                    .thenCompose(cont -> rangeSet.insertRange(store.ensureContextActive(), packOrNull(rangeStart), packOrNull(cont), true)
+                            .thenApply(ignore -> {
+                                if ( quota > 0 ) {
+                                    scanCounterIndexes += recordsScanned.get();
+                                    if (quota <= scanCounterIndexes) {
+                                        return false;
+                                    }
+                                }
+                                return !Objects.equals(cont, rangeEnd);
+                            }));
+        });
+    }
+
+    @Nullable
+    private CompletableFuture<FDBStoredRecord<Message>> deleteIndexIfDangling(FDBRecordStore store, final RecordCursorResult<FDBIndexedRecord<Message>> cursorResult) {
```

Comment:
Hm, the API here is a little weird. It kind of feels like something is quite right about the abstractions, because the whole reason this returns a `CompletableFuture<FDBStoredRecord<Message>>` is so that it can be fed into the indexer code, which expects to be given a record (that it then indexes).

It almost feels like the right abstraction would take a method that then processes each record. In the case of the "scrubIndex" case, it would delete any dangling entries, and in the other cases, it would update the index(es) (conditionally in the "scrubRecords" case).

Reply:
Yeh, it felt a little weird to me too. The way I thought about it is that all the other indexers may return a record to be indexed (and, if applicable, added to the conflict list), but the dangling index is pointing to a null record - therefore, even if dangling, this null record is returned (after adjusting the index entry). 
My other concern was that since there is no record to conflict with, there is a (very) slim chance that a newly written in parallel record may - by an amazing coincidence - have just created this very same index. 
Is it likely to happen before the universe ends? If so, can the index entry itself be added to the conflict list?